alpenhorn/auto_import.py

```diff
@@ -48,52 +56,81 @@ def _import_file(node, root, file_path):
     - The file.
     - Information on the file, if it is of type "corr".
     - Indicates that the file exists on this node.
+
+    Parameters
+    ----------
+    node : storage.StorageNode
+        The node we are processing.
+    file_path : string
+        Path of file on the node to import. If is is an absolute path it must
+        be within the node root, otherwise is is assumed to be relative to
+        the node root.
     """
     # global import_done
     curr_done = True
-    fullpath = os.path.join(root, file_path)
-    log.debug("Considering %s for import." % fullpath)
-
-    # Skip the file if ch_master.py still has a lock on it.
-    dir_name = os.path.dirname(file_path)
-    base_name = os.path.basename(file_path)
-    if os.path.isfile(os.path.join(root, dir_name, ".%s.lock" % base_name)):
-        log.debug("Skipping \"%s\", which is locked by ch_master.py." % fullpath)
+
+    log.debug("Considering \"%s\" for import.", file_path)
+
+    # Occasionally the watchdog sends events on the node root directory itself. Skip these.
+    if file_path == node.root:
+        log.debug("Skipping import request on the node root itself \"%s\"", node.root)
+        return
+
+    # Ensure the path is an absolute path within the node
+    if os.path.isabs(file_path):
+        if not in_directory(file_path, node.root):
+            log.error("File \"%s\" was not an absolute path within the node \"%s\"", file_path, node.root)
+            return
+    else:
+        file_path = os.path.join(node.root, file_path)
+    abspath = os.path.normpath(file_path)
+
+    # Skip requests to import a directory. Again these are occasionally sent by the watchdog
+    if os.path.isdir(file_path):
+        log.debug("Path to import \"%s\" is a directory. Skipping...", file_path)
+        return
+
+    relpath = os.path.relpath(abspath, node.root)
+
+    # Skip the file if there is still a lock on it.
+    dir_name, base_name = os.path.split(abspath)
+    if os.path.isfile(os.path.join(dir_name, ".%s.lock" % base_name)):
+        log.debug("Skipping \"%s\", which is locked by ch_master.py.", file_path)
         return
 
     # Check if we can handle this acquisition, and skip if we can't
-    acq_type_name = ac.AcqType.detect(file_path, node)
+    acq_type_name = ac.AcqType.detect(relpath, node)
     if acq_type_name is None:
-        log.info("Skipping non-acquisition path %s." % file_path)
+        log.info("Skipping non-acquisition path \"%s\".", file_path)
```

Comment:
should this say relpath? At least that's the one you checking in AcqType.detect

Reply:
Hmmmm. That's a good point. I think it probably should stay as `file_path`, basically as it's good to give the original path (before it's been processed internally) so it's easier for the user to figure out what went wrong.